### ISSUE TITLE
Question: Why does image_fail_test_1 output differ when run by ctest vs run directly?

### ISSUE THREAD

**Issue by rouson** [sourceryinstitute/OpenCoarrays#401]
[issue stat img]: https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square

| Avg response time              |
|:------------------------------:|
| ![Issue Stats][issue stat img] |

# Defect/Bug Report #

- OpenCoarrays Version: 1.9.0-4-g6a6e529
- Fortran Compiler: GCC 7.1.0
- C compiler used for building lib: GCC 7.1.0
- Installation method: `install.sh`
- Output of `uname -a`: Linux Sourcery-Linux-VM 4.4.0-77-generic #98-Ubuntu SMP Wed Apr 26 08:34:02 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- MPI library being used: MPICH 3.2
- Machine architecture and number of physical cores: Sourcery Institute virtual machine running on macOS Sierra with 4 cores
- Version of CMake: 3.7.0

## Observed Behavior ##
```bash
$ caf image_fail_test_1.f90 
$ cafrun -n 4 ./a.out
IMAGE FAILED!

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 12315 RUNNING AT Sourcery-Linux-VM
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```
## Expected Behavior ##
I think `Test passed.` should be printed above.

## Steps to Reproduce ##
```bash
git clone https://github.com/sourceryinstitute/opencoarrays
cd opencoarrays
./install.sh --install-prefix /opt/opencoarrays/1.9.0/gnu/7.1.0 --with-mpi /opt/mpich/3.2.0/gnu/7.1.0
source /opt/opencoarrays/1.9.0/gnu/7.1.0
cd src/tests/uit/fail_images
caf image_fail_test_1.f90
cafrun -n 4 ./a.out
```

# Question #
1. Is the above behavior correct?
2. If so, how is it that ctest reports the test as passing?  I see "Test passed" in the text below, but don't understand why it appears when I run the test via ctest but not when I run the same executable program simply with `cafrun` as above?
```bash
$ ctest -R fail_test_1
Test project /home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0
    Start 44: image_fail_test_1
1/1 Test #44: image_fail_test_1 ................   Passed    0.03 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.03 sec
rouson@Sourcery-Linux-VM:~/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0$ ctest -R fail_test_1 --verbose
UpdateCTestConfiguration  from :/home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0/DartConfiguration.tcl
Test project /home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 44
    Start 44: image_fail_test_1

44: Test command: /opt/mpich/3.2.0/gnu/7.1.0/bin/mpiexec "-np" "4" "-disable-auto-cleanup" "/home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/1.9.0/src/tests/unit/fail_images/image_fail_test_1"
44: Test timeout computed to be: 9.99988e+06
44: IMAGE FAILED!
44:  Test passed.
44: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
44: This typically refers to a problem with your application.
44: Please see the FAQ page for debugging suggestions
1/1 Test #44: image_fail_test_1 ................   Passed    0.02 sec

The following tests passed:
	image_fail_test_1

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.02 sec

```

[links]:#
[Google]: https://www.google.com
[Stack Overflow]: https://stackoverflow.com
[fenced codeblocks]: https://help.github.com/articles/creating-and-highlighting-code-blocks/
[GFortran bug tracker]: https://gcc.gnu.org/bugzilla/
[Intel MPI]: https://software.intel.com/en-us/intel-mpi-library
[MPT]: http://www.sgi.com/products/software/sps.html
[MVAPICH]: http://mvapich.cse.ohio-state.edu
[MPICH]: https://www.mpich.org
[Open-MPI]: https://www.open-mpi.org
[CONTRIBUTING.md]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[Helping Out]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md#helping-out
[official release]: https://github.com/sourceryinstitute/OpenCoarrays/releases
[CMake install]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/INSTALL.md#cmake-scripts
[homebrew]: https://brew.sh
[issues page]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[mailing list]: https://groups.google.com/forum/#!forum/opencoarrays
[install.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/install.sh

**Comment by vehre** [comment]
First guess: The flag to not terminate the whole program when one process is failing is missing from the the call to mpirun/mpiexec in caf. On mpirun this is: -disable-auto-cleanup

**Comment by zbeekman** [comment]
Hi Damian,

I think that when you run the test by itself, you need to add `-disable-auto-cleanup` because the MPI runtime needs to know that it is running in fault-tolerant mode. Oh, @vehre just commented to the same effect. Not sure if it will work now, but in the nearly finished `caf` and `cafrun` refactor you will be able to specify a flag for failed images, and other flags will just pass through. So try:

```
cafrun -np 4 -disable-auto-cleanup ./a.out
```

I just double checked the state of `cafrun` and tried this myself, and it works.

There was some extraneous code in CMakeLists.txt that caused a little bit of confusion for me that I have cleaned up... but the main issue seems to be the `-disable-auto-cleanup` flag missing.

**Comment by zbeekman** [comment]
@rouson I think we can close this if you agree... I'll implement a failed-images flag in the revised `cafrun` to help ensure users see that they need to add an extra flag.

**Comment by rouson** [comment]
Let's not close this just yet.  At a minimum, we should document this somewhere prominent (presumably the README.md) if adding the flag is the recommended practice.  Alternatively, let's consider whether to just embedding the flag in cafrun for several reasons:

1. Telling people to explicitly add the flag exposes something MPICH-specific to the user.
2. Failed images are part of the Fortran 2015 standard. Let's make the standard behavior the default behavior.  

One compiler vendor has cost me innumerable hours of my life because I sometimes forget that I need to pass a `-standard-semantics` flag to force the vendor's compiler obey the language standard.

**Comment by zbeekman** [comment]
I guess the real question is as follows: what is the down side, if any of always passing the flag? I think, if I understand correctly---which I may not, that passing the flag can cause problems when user code is *NOT* using failed images explicitly. Certainly more investigation is necessary, and at a minimum a flag needs to be added to `cafrun`

**Comment by rouson** [comment]
Let's keep this open until we add a new non-MPICH-specific flag such as
```bash
cafrun -np 4 -allow-failed-images ./a.out
```

**Comment by zbeekman** [comment]
Investigate possible drawbacks of passing `--disable-auto-cleanup`

**Comment by jeffhammond** [comment]
The default behavior should align with the common practice of Fortran application programmers.  Are more than half of all Fortran coarray programs explicitly fault-tolerant or checking for image failure and ?  If not, then the principle of least surprise says that you should not make `--disable-auto-cleanup` the default.

**Comment by rouson** [comment]
@jeffhammond you make a valid suggestion.  Hopefully you saw my comment up-thread, which provides some personal experience as a counterbalance.  Given how reluctant the Fortran community can be to adopt new language features, I'd like apply the principle of least surprise to those trying new features unless doing so breaks existing code.  So far, we're not aware of any way in which this flag would do so. 

Sadly, I'm realizing I should have written "at least three compilers" in my previous comment rather than "one compiler vendor."   It used to be the case that the Cray, Intel, and Portland Group compilers all required a special flag for the automatic array reallocation that the standard requires for the left-hand side of the following assignment:
```fortran
integer, allocatable :: array(:)
array = [ 0, 1, 2]
```
I'm confident that I've lost a cumulative total of multiple days of my life to searching for coding errors that turned out to not be coding errors and were instead instances of one of the aforementioned compilers exhibiting non-standard behavior by default in the above array assignment.  Fortunately, all three have dropped the flag, partly due to my begging them in issue reports.
 
Zaak just emailed the MPICH discussion mailing list for feedback. Let's see what they say before deciding, but I hope we'll set a high bar for making a decision to violate the standard by default.  Failed images are in the draft of the standard so I hope we'll support failed images by default.

**Comment by zbeekman** [comment]
> If not, then the principle of least surprise says that you should not make --disable-auto-cleanup the default.

Well, my question to the MPICH list was to better understand what surprises may accompany passing `--disable-auto-cleanup` to someone who is not expecting it, if any. As of right now it looks like there are either MPI standard bugs, and/or library implementation bugs that causes hanging and other undesirable behavior if the flag is passed and an error is encountered, but users are not explicitly using fault tolerance... As Damian pointed out, not being standard compliant, or requiring a special runtime flag to conform to standard semantics is fairly surprising. So I want to understand, what are the surprises that may come with `--disable-auto-cleanup` so I can make an educated assessment of what is least surprising.

**Comment by jeffhammond** [comment]
Not being standard-compliant is never surprising.  How many truly standard-compliant programming systems can you name?

The MPI world does well because there are really only three unique implementations that support more than one platform and many others derive their compliance from one of these (MPICH).

There might be three OpenMP 4.5 implementations by the end of this year, if you count ~30x slower than serial as a correct implementation of ordered for.

For ISO languages, there are:
- three or four C++14 compilers.
- zero C11 compilers unless you replace glibc with musl.
- two Fortran 2008 compilers.

And note that in many cases, a special flag is required to achieve the latest standard-compliance.  One must also ignore all of the bugs, of course, and there are many of these that are show-stoppers, not just nits.

While I admire OpenCoarrays for trying, given that GCC doesn't even claim ISO Fortran 2003 compliance, I don't know why any educated user would be surprised that ISO Fortran 2015 coarray fault-tolerance isn't enabled by default.

# References
- http://fortranwiki.org/fortran/show/Fortran+2003+status
- http://fortranwiki.org/fortran/show/Fortran+2008+status
- https://gcc.gnu.org/wiki/Fortran2003Status
- https://gcc.gnu.org/wiki/Fortran2008Status
- http://en.cppreference.com/w/cpp/compiler_support
- https://gcc.gnu.org/wiki/C11Status
- https://software.intel.com/en-us/articles/c11-support-in-intel-c-compiler
- http://www.openmp.org/resources/openmp-compilers/

**Comment by zbeekman** [comment]
@jeffhammond Jeff, all valid points, but the essence of my question remains: What are the draw backs, if any, of `--disable-auto-cleanup` and how does passing it at runtime affect the semantics of a users code?

I'm happy to be pointed to references, but I'm not sure where to start answering this question.

If there are no surprises if it's passed by default then there's no harm, and no reason not to... if there are surprises, other than bugs in implementations, I want to learn about them. I want to know about the bugs too, of course.

**Comment by rouson** [comment]
@jeffhammond 

In broad strokes, you make good points and I'm surprised to hear about the status of C++ because some of the people who are the most ardent C++ proponents love to bludgeon Fortran programmers with the admittedly valid claim that the compiler teams have been too slow to catch up with the standards.  On that note, however, I do have one nit that motivates the following:

** Begin rant **

The cited Fortran 2003 status page frustrates me to no end.  It is perennially out-of-date for multiple compilers.   I inserted a proposal to delete it a year or so ago and I'm strongly considering doing so now.  Please note the following text from the Fortran 2008 standard:

"Any standard-conforming Fortran 2003 program remains standard-conforming under this part of ISO/IEC 1539."

As such, I believe the hassle of maintaining separate 2003 and 2008 pages outweighs the benefit.  Nonetheless, I just  updated the 2003 page to show that there is only one remaining unimplemented Fortran 2003 feature in gfortran 7.2.   And I am very pleased to report that just this morning, I received an email from a gfortran developer who reports having finished implementing that one missing feature.  And the picture is even better than that.  Assuming all the various patches/branches/forks of which I'm aware are accepted into the GCC trunk, I expect that next year's release of GCC 8 will have at least partial support for all major features of **Fortran 2015** by my own subjective definition of "major".

Compiler support for modern Fortran has advanced by leaps and bounds in recent years and I only wish there would be a bit of a shakeout such that the compilers that still lag the standards by multiple versions would fall into disuse or at least that someone would contractually hold their developers' feet to the fire to force them to catch up.  We may not be where we need to be, but we are far beyond where we were even as recently as a year ago.  

** End rant **

The above rant does not negate your main point: every programming system has bugs and most have missing or partially implemented standard features so none are truly standard-compliant.  Let's see what Zaak finds out and go from there.

**Comment by jeffhammond** [comment]
@rouson Having GCC Fortran at the bleeding edge of the standard is the right way to move the community forward.  That vendors were historically (far) ahead of GCC w.r.t. Fortran allowed vendors to pick and choose what they wanted to support and made life difficulty on developers, particularly the financially limited ones (especially grad students!).  As you note, the aggressive adoption of C++ features in GCC *and* Clang puts the pressure on other implementers to keep up both w.r.t. standard-compliance and performance.  It also means that nearly all platforms support three different compilers, which benefits users in many ways.

Once GCC is done with Fortran 2015, perhaps efforts can be redirected to making it performance-competitive with Intel and Cray 😁  Part of this would include optimizing GOMP, assuming it still does an O(n) barrier.

GCC and Clang are very aggressive about implementing the latest C++ standard and have had C++11 and C++14 support for a while.  The biggest issue I see w.r.t. C++17 is that GCC's implementation of PSTL isn't fully baked and Clang is still discussing how to implement PSTL, although one can argue that, since PSTL is a library, either can use [the HPX implementation of PSTL](http://stellar.cct.lsu.edu/pubs/ParallelizingSTL.pdf) (I have not tested this hypothesis).  On the other hand, Intel has a very good implementation of PSTL already, but doesn't claim full support for the language features [here](https://software.intel.com/en-us/articles/c17-features-supported-by-intel-c-compiler).

**Comment by rouson** [comment]
@jeffhammond I misread "C11" as "C++11" in a bulleted item you posted.  With that correction in my reading, the C++ situation matches my impression of the state of affairs.  I didn't realize compiler support for the C standard had fallen so far behind. 

The news on PSTL is encouraging.  I'll be especially interested to see if UPC++ effort moves into the standard namespace.  I frequently tout the fact that Fortran is the only mainstream language with a parallel programming model that works in distributed memory.  That will likely remain true unless and until UPC++ (or possibly but less likely Coarray C++) moves into the C++ standard.

**Comment by jeffhammond** [comment]
@rouson The *compiler* support for C11 is fine in GCC, Clang, ICC, probably Cray, and perhaps others (I've only tested atomics and threads).  The problem is *library* support: glibc doesn't consider C11 threads a priority.  See https://sourceware.org/bugzilla/show_bug.cgi?id=14092 and https://stackoverflow.com/questions/24557728/does-any-c-library-implement-c11-threads-for-gnu-linux for details.

Granted, C11 threads are technically optional, but the proper way to query this doesn't always work (see https://bugs.llvm.org/show_bug.cgi?id=32377, https://software.intel.com/en-us/forums/intel-c-compiler/topic/721059) and the GCC folks insist that implementing preprocessor symbols incorrectly is okay if some websites has details (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58016).

> UPC++ effort moves into the standard namespace

Don't hold your breath.  C++ will probably standardize a sockets-like API in the networking TS, which is something that will be useful to many, whereas UPC++ already works with any C++11 compiler because it doesn't require compiler support and is likely to be used by significantly fewer people.

We wrote [this paper](https://doi.org/10.1109/ICPP.2016.28) to demonstrate that any C++ decent C++ code can be extended to support PGAS features using a library solution.  I don't see a compelling argument to put them in the standard, given that the implementations are unlikely to be portable, in contrast to much of the rest of the C++ libraries.

**Comment by zbeekman** [comment]
>Hi Zaak,
> 
> I'll try my best to explain here. There are a few things to consider.
> 
> 1. Hydra: -disable-auto-cleanup means if an MPI process dies, let other
> processes in the job continue running. Since Hydra (mpiexec) is already
> monitoring MPI processes to detect when one dies, there is no impact
> inside Hydra from passing this option.
> 
> 2. Application behavior: Since the default error handler in MPI is
> MPI_ERRORS_ARE_FATAL, some applications may rely on that fact and expect
> a running job to be aborted/cleaned up if a process quits. With
> -disable-auto-cleanup this will no longer be the case. An application
> can call MPI_Abort() to force the old behavior, however.
> 
> Ken

@afanfa @vehre Do you know:

 1. Is there a way to detect if an image has failed/errored/halted from a different remote image __*without*__ support for user level failure mitigation (ULFM) in the MPI library? I think the answer to this is "*no*," and just want to confirm.
 2. If OpenCoarrays __*is*__ built with failed images support, but the user does not explicitly check for `stat_failed_image` or `stat_stopped_image` *BUT* there is in fact a failed or stopped image, do we explicitly call `MPI_abort` from any image control statements or when a "get" (remote coarray RHS) is encountered? Does the standard specify anything about what the correct course of action is for this situation?

I think that the answers to these two questions will dictate whether we pass `--disable-auto-cleanup` by default. If the answer to 1 is "no" then we certainly don't want to pass the flag if OpenCoarrays is *NOT* built with failed images support. If OpenCoarrays is built with failed image support, but the user is not checking for `stat_stopped_image` or `stat_failed_image` and the library isn't doing any checking for these conditions on it's own, followed by a subsequent call to MPI abort, then I think we DO *NOT* want to pass `--disable-auto-cleanup` by default because user code will just hang indefinitely if it reaches a synchronization point and there is a stopped/failed image.

Thoughts, @rouson @afanfa @vehre ?

**Comment by afanfa** [comment]
The answer to 1 is no. The default for the MPI implementations without fault tolerant support is aborting the application.
The answer to 2 is that we call MPI_Abort only for control statements without a stat= attribute. Gets and Puts from/to a failed image is comparable to read/write on /dev/null. If the user does not pass any stat= attribute for gets or puts the application should keep going.

**Comment by vehre** [comment]
to question 1.: Correct, the answer is 'no'.
to answer q 2: The user does not need to check for `stat_failed_image` or `stat_stopped_image` but needs to provide the STAT=variable to all major coarray-statements, i.e., to all that support a STAT= (get, send, getsend, allocate, ...) When the user does not provide the STAT= and a failed image is present the program is aborted. This is how I remember what the standard says.

Albeit I am not sure, on the what the exact condition is as stated by the standard, when no STAT= is given. It could be that:
1. the program is aborted when any failed/stopped image is present, or
2. when the image to communicate with is failed/stopped.

When I remember correctly, then "1. the program is aborted when any failed/stopped image is present " is implemented in opencoarrays. I can't find at the moment in the standard where this is mandated for get, send and sendget, but F2015 11.6.11 § 9 says: `If an error condition occurs during execution of an image control statement that does not contain the STAT= specifier in a sync-stat, error termination is initiated.` (I know its for SYNC, but I am positive similar statements are made for get and consorts).

**Comment by zbeekman** [comment]
@afanfa & @vehre thanks so much for the response.

It seems to me, then, that if OpenCoarrays is built with failed image support that passing `--disable-auto-cleanup` by default should be *relatively* safe; hopefully any gets or puts won't hang (and will act like reading/writing to `/dev/null` and then the non-failed images will be able to progress to the next image control statement where one of two things will happen:

 1. They provide `stat=` and it is their responsibility to handle error termination, or
 2. They don't provide `stat=` and in the image control statement the library initializes error termination and at some point calls `MPI_Abort`.

So, if this is the case, then I think it is safe to go with @rouson's suggestion that we pass `--disable-auto-cleanup` by default if the library was built with failed image support (meaning that the underlying MPI has ULFM capabilities. Great, thanks for clearing this up for me!

**Comment by jeffhammond** [comment]
Is there a way to associated the with building for 2008 or 2015 or provide a range of methods (OpenCoarrays build time, user code compile time, job launch option, runtime env var) that allow me to enable MPI auto-cleanup?  I am currently writing 2008 coarrays because that is what is portable.  I cannot use 2015 features in these programs, and those features include the functions that check for failed images.

**Comment by zbeekman** [comment]
@jeffhammond Assuming you're asking specifically about OpenCoarrays + GFortran....

MPI auto-cleanup will always be available at runtime, no matter what. Additional/non-specialized arguments to the `cafrun` wrapper script get passed through to `mpiexec`. Currently, as of OpenCoarrays 1.9.1, you must explicitly pass `--disable-auto-cleanup` to `cafrun` or `mpiexec` to turn it off. It's the default behavior of MPI, and up until this point the default behavior of the `cafrun` script too.

I am nearly done with a major cleanup and refactoring of the compiler wrapper script, `caf` and job launch wrapper script `cafrun`, and the discussion here has been about the default behavior for `cafrun`. The only purpose of `cafrun` is really to try to abstract job launching away from MPI, so that in theory you could have any number of back ends, like the experimental GASNet back end and the experimental OpenSHMEM back end. *We will always provide a flag to re-enable MPI auto-cleanup at runtime, no matter what the default is.* Moving forwards the default behavior of `cafrun` will be:

 - to disable MPI auto-cleanup if:
    - MPI ULFM support was detected at build time for OpenCoarrays (i.e. MPICH >= 3.2) and the library (>= 1.9.0) and GFortran compiler (i.e. >= 7) have failed image support.
    - The user __*did not*__ specifically request that failed image support be turned off via the `-DCAF_ENABLE_FAILED_IMAGES:BOOL=FALSE` flag passed to CMake when building the library, due to say, [a bug in MPICH 3.2](https://github.com/sourceryinstitute/OpenCoarrays/issues/439).
    - The user did not pass a flag explicitly to `cafrun` to request it be re-enabled

So to summarize:

> OpenCoarrays build time

MPI auto-cleanup defaults to OFF, only when: OC >= 1.9.0 && MPI implementation supports ULFM (i.e. MPICH 3.2) && GFortran supports failed images (GCC >= 7)

> user code compile time

Whether MPI auto-cleanup is off or on, a call to MPI_Abort will be made if `stat=...` is not passed to image control statements. This should mostly mimic MPI auto-cleanup behavior

> job launch option

Quite a few conditions must be met to *disable* MPI auto-cleanup by default, MPI lib support must be detected, GFortran support must be detected, and support for failed images can be explicitly turned of at compile time of OC by the users. At runtime, there will always be an option to re-enable MPI auto-cleanup:

 - Pass the corresponding flag to `cafrun`, we haven't settled on what it will be called yet...
 - If the semantics of `mpiexec` support overriding early flags with late flags, passing `--enable-auto-cleanup` to `cafrun` will get passed through to `mpiexec` and should have the desired effect
 - Launching the job directly with `mpiexec` rather than `cafrun` will default to MPI auto-cleanup

> runtime env var

No planned support for this, at this time, but would be fairly trivial to implement, since it is only enabled or disabled at runtime (call to `mpiexec`) any way.

phew, sorry this is so long, but I hope I've clarified and addressed all relevant points

**Comment by zbeekman** [comment]
@jeffhammond TL;DR of my last comment is that:

1. MPI auto-cleanup is only ever disabled at runtime, and no matter what the default is, should be able to be re-enabled trivially. Full support for failed images must be detected to make it the default
2. If you build OC supporting failed images, you shouldn't have to change your code, the semantics *should* be the same, or very similar, even when MPI auto-cleanup is disabled, because OC calls MPI Abort on stopped/failed images, which mimics the auto-cleanup behavior.

**Comment by zbeekman** [comment]
Resolved in PR #440. Default behavior is to disable auto-cleanup if failed image support/ULFM is available at build time. This behavior can be over-ridden with `--reenable-auto-cleanup` passed to `cafrun` at runtime. If MPI/GFortran can't support failed images, then this flag is a noop and auto-cleanup is always on, unless the `--disable-auto-cleanup` is passed to `mpiexec` or `cafrun`.

When auto cleanup is disabled, user code should behave as expected, and it will abort if stopped or failed images are detected at each image control statement without a user supplied `stat=`.
